sdk/src/main/java/com/google/cloud/dataflow/sdk/util/common/Counter.java

```diff
@@ -418,17 +452,191 @@ public boolean isCompatibleWith(Counter<?> that) {
 
   //////////////////////////////////////////////////////////////////////////////
 
-  /** The name of this counter. */
-  protected final String name;
-
   /** The kind of aggregation function to apply to this counter. */
   protected final AggregationKind kind;
 
   /** The commit state of this counter. **/
   protected final AtomicReference<CommitState> commitState;
 
+  /** which name to use from the original context. */
+  public static enum NameType {
+    /** Counter does not have a structured name. */
+    UNSTRUCTURED,
+
+    /** Structured name should use the original name. */
+    ORIGINAL,
+
+    /** Structured name should use the optimized name. */
+    OPTIMIZED
+  }
+
+  /** A hashable name structure which will uniquely identify a counter. */
+  public interface Name {
+    public String getName();
+    public NameContext getContext();
+    public NameType getType();
+  }
+
+  /** A StructuredName which uses the Original step name from the NameContext. */
+  public static class StructuredNameOriginal implements Name {
+
+    /** The name of this counter. */
+    private final String name;
```

Comment:
I'm not sure AutoValue adds much to these classes because they have to do custom hashCode / Equals logic to only compare parts of the name context which make sense for them. Thoughts?